src/backend/src/signer.rs

```diff
@@ -151,7 +159,52 @@ pub async fn btc_principal_to_p2wpkh_address(
 ///
 /// # Errors
 /// Errors are enumerated by: `TopUpCyclesLedgerError`
-#[allow(clippy::unused_async)] // TODO: Remove once the code is implemented
 pub async fn top_up_cycles_ledger(request: TopUpCyclesLedgerRequest) -> TopUpCyclesLedgerResult {
-    todo!("Add code that tops up the cycles ledger per this request: {request:?}")
+    let cycles_ledger = CyclesLedgerService(*CYCLES_LEDGER);
+    let account = Account {
+        owner: ic_cdk::id(),
+        subaccount: None,
+    };
+    let (ledger_balance,): (Nat,) = cycles_ledger
+        .icrc_1_balance_of(&account)
+        .await
+        .map_err(|_| TopUpCyclesLedgerError::CouldNotGetBalanceFromCyclesLedger)?;
+
+    if ledger_balance < request.threshold() {
+        // Decide how many cycles to keep and how many to send to the cycles ledger.
+        let own_canister_cycle_balance = Nat::from(ic_cdk::api::canister_balance128());
+        let to_send = own_canister_cycle_balance.clone() / Nat::from(100u32)
+            * Nat::from(request.percentage());
+        let to_retain = own_canister_cycle_balance.clone() - to_send.clone();
+
+        // Top up the cycles ledger.
+        let arg = DepositArgs {
+            to: account,
+            memo: None,
```

Comment:
Maybe set `created_at_time` as well?

Reply:
I don't believe that will be needed.  Or what am I missing?  This is used immediately.

Reply:
The `created_at_time` is used for the [deduplication](https://github.com/dfinity/ICRC-1/tree/main/standards/ICRC-1#transaction-deduplication-) mechanism of an icrc1 ledger. That's why I end generally always setting it. But indeed maybe not necessary here.

Reply:
If neither memo nor time is set, a standards compliant ledger should not deduplicate, IIRC.  Indeed:

> If the client did not set the created_at_time field, the ledger SHOULD NOT deduplicate the transaction.

So I think we are fine.